### ISSUE TITLE
Choice to run in in `--enable-asserts` mode

### ISSUE THREAD

**Issue by zoechi** [dart-lang/dart-pad#310]
It would be great to be able to switch between checked/production mode for "run" to be able to demonstrate the different behavior for example for StackOverflow questions.

**Comment by lukechurch** [comment]
I'll add this to the service layer, should be an easy addition.

**Comment by kwalrath** [comment]
:+1: I was just running examples from the language tour, and realized that all the asserts were doing nothing.

**Comment by kwalrath** [comment]
I'd like this reopened, since there's still no UI (that I can find) for selecting checked mode.

Or is there, @devoncarew?

**Comment by devoncarew** [comment]
We could just default to checked? Right now we use dart2js' default, which is production mode.

**Comment by kwalrath** [comment]
@zoechi what do you think? For me, checked > unchecked, but it might be nice to toggle between the two (perhaps using a URL option, so it's not yet another thing onscreen).

**Comment by zoechi** [comment]
There is also the upcoming strong mode. 
It would be great to be able to demonstrate different behavior and error messages resulting from different modes.

**Comment by devoncarew** [comment]
I think we should choose a default for dartpad-using-dart2js; probably checked mode.

There's another issue tracking experimenting with using ddc in dartpad. In that dartpad, we'd turn on strong mode in the analyzer. It _may_ make sense to have a UI option to switch between dart2js and ddc then, at least for a little while.

**Comment by srawlins** [comment]
devoncarew@ is referring to #694.

**Comment by zoechi** [comment]
I think this is obsolete

**Comment by kwalrath** [comment]
I don't think this issue is obsolete, because we still need a way to say "honor assert statements". Most of our language/library tour examples use asserts right now. I guess we could change them to prints, but that's a lot of code to change... And pedagogically it'd be good for people to be able to easily see what assert does. (This assumes that we'll still support asserts and have some sort of developer mode option... that was true last I heard.)

/cc @chalin

**Comment by srawlins** [comment]
I'll rename/repurpose the issue to add an `--enable-asserts` flag.

**Comment by kwalrath** [comment]
Actually, I think asserts are enabled all the time now. E.g. this code works as I'd expect it to:

```
void main() {
  for (int i = 0; i < 5; i++) {
    print('hello ${i + 1}');
    assert(i == 0, "i is $i");
  }
}
```

So I'm happy. :)

**Comment by kwalrath** [comment]
@zoechi feel free to reopen if you don't want asserts to work all the time... I prefer having asserts on by default, but I don't object to having a way to turn them off, as long as it doesn't complicate the UI.

**Comment by sergeytsivin** [comment]
As of now, assert statements are ignored in dartpad: https://dartpad.dartlang.org/

**Comment by kwalrath** [comment]
> As of now, assert statements are ignored in dartpad: https://dartpad.dartlang.org/
> 
> Should this issue be reopened?

Thanks for letting us know. I'm reopening it.

**Comment by chalin** [comment]
Enable-asserts was the default for DartPad but only for a brief moment in time? Why was it removed/disabled? Could it be re-enabled again? 😄

**Comment by filiph** [comment]
This tripped me up today. I tried to build a simple example that lets a learner try to fix a piece of code that would throw an assertion error in real life. But it ran in DartPad just fine, and I didn't know why.

Just thought I'd mention it, since the following seems like an obvious minimalist codelab thing to do:

```dart
int fib(int n) {
  // Fix this!!
  return 0;
}

void main() {
  assert(fib(1) == 1);
  assert(fib(5) == 5);
  assert(fib(10) == 55);
}
```

Not saying this is super important, just surprising.

**Comment by chalin** [comment]
One of the main uses cases for DartPad, IMHO, is in support of teaching Dart via codelabs and tutorials, for example. Now if a behavior of DartPad surprised a very seasoned Dart developer like you @filiph, it has the potential for causing much more surprise and grief to Dart newcomers. In that sense, I'd say that this is an important issue to address.

**Comment by devoncarew** [comment]
Instead of having an opt-in / out for this, I wonder if we shouldn't just enable assertions for all dartpad snippets?

That would be a bit challenging for flutter samples, as the flutter framework has a fair number of assertions for debug-time - they would definitely slow down a release mode app.

**Comment by sigmundch** [comment]
@devoncarew I was also wondering about the heavy use of assertions in flutter code. In particular:

* how often are dartpad flutter samples used to showcase the production quality of flutter? If that's common, enabling assertions will likely give the wrong impression
* similarly are there any worries that some users would be concerned if they see the unoptimized output of dart2js?

If these indirect impressions are important from DartPad's perspective, I'd be inclined to have a toggle to enable/disable them. If not, I'd be fine enabling them by default.

**Comment by jpangburn** [comment]
I am new to Dart and was working my way through the language tour, and was doing some tests with String interpolation and assert.  Became suspicious that asserts weren't working and was disappointed to find that stuff that clearly should fail does not fail.

There are some comments above about who's using DartPad and what their priorities are.  As a new user (who came to Dart because of Flutter), I have the following perspective:
1.  People new to Flutter will likely be running Flutter code in either VSCode (my case because ios) or Android Studio.  Only after reading some comments here did it occur to me that people might use DartPad for Flutter, but I don't know enough to understand why or how.
2. People new to Flutter will want to learn Dart basics somewhere other than Flutter, because while hot reload is fast- there's nothing like the simplicity of DartPad to try some core language stuff like string interpolation, fiddling with numbers, assertions, classes, etc. without the overhead of a GUI.

So having DartPad be optimized for production doesn't make a lot of sense to me.  I'm sure more seasoned users have excellent reasons for this, but would they still even be using DartPad for anything beyond quick language tests?  Or would they be writing Dart tests in the context of their larger application and thus be running tests from the command line or something?

Anyway, hope this perspective helps.  After I hit the Comment button, I'll be reading about how to run Dart (without Flutter) in VSCode so I can continue the language tour in an environment where failed assertions actually cause an error.  DartPad is apparently not intended for my current use case.
